### DISCUSSION TITLE
Is it still feasible to direct an internal link to a page that has not been created yet, without specifying the page number ?

### DISCUSSION THREAD

**Discussion by nsimonovici** [py-pdf/fpdf2#729] [Q&A] [answered]
Hi there,

As I switch from fpdf to fpdf2, I notice a change in behaviour in my internal links.

I was using the "add_link" to set up internal links, making the cell clickable, while saving the created link (result out of "add_link") in a list.

As my layout can move depending of pictures or texts sizes, I did not want to hardcode the page to where links are directing.
That's why I would trigger the "set_link" method of desired link in the list at the moment of my choice when building my PDF file, directing automatically the priorly created link in cell to a page created afterwards.

But that was before. Now I can read from the latest release that :
![image](https://user-images.githubusercontent.com/44460830/227258146-bd0ad4db-1ce2-4709-8358-fb94fac8bd98.png)

And my "trick" does not work anymore, sadly.

Therefore is there a retro compatibility way to achieve what was working perfectly in fpdf ? Maybe downgrading my package might by a solution. EDIT : it is not (tried version 2.6.0)

Thanks for any kind of assistance

**Comment by Lucas-C** [comment, marked as answer]
Hi!

Thank you for reaching out @nsimonovici and welcome to the `fpdf2` community 😊

What you describe is very strange and may be a bug...
`FPDF.add_link()` in the latest releases should be fully backward compatible with the earlier versions of `fpdf2` and even `PyFPDF`.

We only added some extra features to it in `v2.6.1`.
Calling `set_link()` is simply now **optional**: whereas you previously had to call `link = pdf.add_link()` **and** `pdf.set_link(link, page=2)`, you can now just make a single call `link = pdf.add_link(page=2)`, if you know the target page number upfront.
But calling `set_link()` is still valid.

We have a set of unit tests regarding `add_link()` that ensure the same features still work as they used to:
https://github.com/PyFPDF/fpdf2/blob/master/test/test_links.py
And those tests are all passing.

Could you please provide some minimal Python code reproducing your issue?
If you don't know how to build a minimal reproducible example, check this tutorial: https://stackoverflow.com/help/minimal-reproducible-example

**Reply by nsimonovici** [reply]
Hi Lucas !

Thank you for answering so quickly.

Here's some code reproducing my issue. I installed fpdf2 v2.6.1.

```python
# coding:utf-8

import os
from fpdf import FPDF
from fpdf.enums import XPos, YPos

pdf = FPDF()
pdf.set_font("helvetica")

pdf.add_page() # page 1

linktopage2 = pdf.add_link(page=2)
pdf.cell(txt='This is an example of internal link to page 2', link=linktopage2, new_x=XPos.LMARGIN, new_y=YPos.NEXT)

linktounknownpage = pdf.add_link()
pdf.cell(txt='This is an example of internal link to a yet to determine page', link=linktounknownpage)

pdf.add_page() # page 2

pdf.set_link(linktounknownpage)
pdf.cell(txt='This should return back to page 1', link=linktounknownpage)

pdf.output(os.path.join(os.getcwd(), "test_links.pdf"))
```

Let me know what I'm doing wrong x) this is most likely a newbie issue

**Reply by Lucas-C** [reply]
Oh OK, I see the problem 😅

When you call `pdf.set_link(linktounknownpage)`, it binds `linktounknownpage` to the current page, which is the **second page**, as you perform this call after `pdf.add_page() # page 2`.

You want either to:
* perform the call to `pdf.set_link(linktounknownpage)` **BEFORE** the call to `pdf.add_page() # page 2`
* pass `page=1`: `pdf.set_link(linktounknownpage, page=1)`

**Reply by nsimonovici** [reply]
I understand what you mean. But wouldn't it require for me to know exactly what page I want as a destination ? In advance ?

Let's say I write a table of content in page 1. I want the titles to direct to the appropriate pages later on in the document.

I would call "add_link" in page 1 (clickable table of content), and because I build lots of stuff between this call of "add_link" and the appropriate page related to this particular title, I don't know in advance what page this will be.
Therefore how can I set the link to the related page later on in the code ? When I actually create the cell corresponding to the title in table of content ?

I sincerely apologize if my explanations are not clear enough.

**Reply by Lucas-C** [reply]
You are right, there is indeed a bug!
I'm working on a fix right now.
In the meantime, have you considered using `FPDF` methods dedicated to building a document outline & table of contents?
_cf._ https://pyfpdf.github.io/fpdf2/DocumentOutlineAndTableOfContents.html

**Reply by nsimonovici** [reply]
Thanks Lucas !

I saw the dedicated methods to creating outline and toc but as I already finished my project when I moved from FPDF to FPDF2 I was lazy and kept my home-made outline.

But depending on how a fix for the bug would be available I might in the end use the dedicated method available in FPDF2.

**Reply by Lucas-C** [reply]
A fix is coming in this PR: https://github.com/PyFPDF/fpdf2/pull/730

You can test it using `pip`:

    pip install git+https://github.com/PyFPDF/fpdf2.git@master

**Reply by nsimonovici** [reply]
I can confirm it is now working as expected ! Kudos !

Any idea when would that be published in an official version ?

**Reply by Lucas-C** [reply]
> Any idea when would that be published in an official version ?

I'm planning on releasing a new version next week.

**Reply by Lucas-C** [reply]
A fix has been released in v2.7.0: https://github.com/PyFPDF/fpdf2/releases/tag/2.7.0

**Reply by nsimonovici** [reply]
Thanks Lucas you're amazing, that was handled smoothly.

All the best !

**Comment by osresearch** [comment]
I don't know if this is a regression or expected behaviour, but I ran into a problem upgrading to fpdf2 with the same error.  Pre-building all of the links by calling `add_link` *before* any calls to `add_page` results in a failure.  The test case in https://github.com/PyFPDF/fpdf2/blob/master/test/test_links.py adds the page before generating the forward link to section 1, so it doesn't trigger it this behaviour, but this minor tweak to the test fails:

```python
from fpdf import FPDF
pdf = FPDF()
pdf.set_font("helvetica")

link_to_section1 = pdf.add_link()  # *before* add_page

pdf.add_page()  # page 1
pdf.cell(text="Section 1", link=link_to_section1)

pdf.add_page()  # page 2
pdf.set_link(link_to_section1, page=pdf.page)
pdf.cell(text="Section 1: Bla bla bla")
```

results in:

```
ValueError: Cannot insert link 1 with no page number assigned
```

**Reply by Lucas-C** [reply]
Hi @nsimonovici 🙂

Given that there is an explicit error explaining the problem, yes this is the expected behaviour.

If you have a specific use case in mind where you would like to create links **BEFORE** calling `.add_page()`,
and you can describe it to us and explain why calling `.add_page()` before is not possible,
we could consider supporting this use case in `fpdf2` 🙂

**Reply by andersonhc** [reply]
`add_link()` creates an initial object pointing to the current page - 0 in your case - and using the link in `cell()` without a valid page is raising the exception.

To make it work you can:
- Create the links after the first call to `add_page()`
- Create all the links pointing to the page 1 with `link_to_section1 = pdf.add_link(page=1)` - It will initially point to page 1 but update when you call `set_link()`

I will wait for @Lucas-C input on how we can improve this behavior on fpdf2. I don't think only removing the validation on `cell()` is a good idea because we can end up with an invalid document if you never call `set_link()` to point to a valid page.
I believe the best approach is to change [`add_link()` and default to 1 if self.page = 0](https://github.com/py-pdf/fpdf2/blob/bb3881b4d3de123967286fb1185e8a6424e903c9/fpdf/fpdf.py#L2226) since a page 1 is mandatory and will be automatically added at `output()` if it hasn't been done before.

**Reply by osresearch** [reply]
![calendar on an eink tablet](https://github.com/osresearch/pdfcal/raw/main/tablet.jpg)

In the [pdfcal](https://github.com/osresearch/pdfcal) code it was not a problem to move the link pre-generation for everyday of the year until the title page was generated, although it used to work in fpdf 1, which is why I was wondering if it was a regression.

Perhaps a way to make the error clearer is to have `pdf.add_link()` throw an error if it is called too soon without an explicit page number; having the error occur when the link is referenced happens quite a bit later rather than on the line that actually needs to be fixed.

**Reply by Lucas-C** [reply]
> I believe the best approach is to change [`add_link()` and default to 1 if self.page = 0](https://github.com/py-pdf/fpdf2/blob/bb3881b4d3de123967286fb1185e8a6424e903c9/fpdf/fpdf.py#L2226) since a page 1 is mandatory and will be automatically added at `output()` if it hasn't been done before.

If we change the default value, we won't be able to distinguish the case where an explicit value is provided, and when we should use the current page number. Or did I miss something?

> Perhaps a way to make the error clearer is to have pdf.add_link() throw an error if it is called too soon without an explicit page number; having the error occur when the link is referenced happens quite a bit later rather than on the line that actually needs to be fixed.

Sounds like a good idea 🙂
Would you like to submit a Pull Request?

**Reply by osresearch** [reply]
I agree with @andersonhc suggestion of setting the default to page 1 (iff no pages have been added yet) rather than throwing an error.  I've made the change in https://github.com/py-pdf/fpdf2/pull/1337/
